File: longest-repeating-character-replacement/Solution424.java
// https://leetcode.com/problems/longest-repeating-character-replacement/

import java.util.HashMap;

class Solution424 {

    // T: O(n), S: O(1)
    public int characterReplacement(String s, int k) {
        int result = 0, left = 0, mostFreqCount = 0;
        var counts = new HashMap<Character, Integer>();
        // Iterate through the string with a sliding window
        for (int right = 0; right < s.length(); right++) {
            // Update the frequency of the current character
            char rightChar = s.charAt(right);
            counts.put(rightChar, counts.getOrDefault(rightChar, 0) + 1);
            // Track the maximum frequency of any character in the current window
            mostFreqCount = Math.max(mostFreqCount, counts.get(rightChar));
            // If the number of replacements needed (window size - max frequency) exceeds k,
            // shrink the window
            int replacementsNeeded = right - left + 1 - mostFreqCount;
            if (replacementsNeeded > k) {
                char leftChar = s.charAt(left);
                counts.put(leftChar, counts.get(leftChar) - 1);
                left++;
            }
            result = Math.max(result, right - left + 1);
        }
        return result;
    }

    public static void main(String[] args) {
        var sol = new Solution424();
        var s1 = "ABAB";
        var s2 = "AABABBA";
        System.out.println(sol.characterReplacement(s1, 2)); // 4
        System.out.println(sol.characterReplacement(s2, 1)); // 4
    }

}

File: longest-repeating-character-replacement/solution424.py
# https://leetcode.com/problems/longest-repeating-character-replacement/

class Solution424:

    # T: O(n), S: O(1)
    def characterReplacement(self, s: str, k: int) -> int:
        result = 0
        left = 0
        most_freq_count = 0
        counts = {}

        for right in range(len(s)):
            right_char = s[right]
            counts[right_char] = counts.get(right_char, 0) + 1
            most_freq_count = max(most_freq_count, counts[right_char])

            replacements_needed = right - left + 1 - most_freq_count
            if replacements_needed > k:
                left_char = s[left]
                counts[left_char] -= 1
                left += 1

            result = max(result, right - left + 1)

        return result


def main() -> None:
    sol = Solution424()
    s1 = "ABAB"
    s2 = "AABABBA"
    print(sol.characterReplacement(s1, 2))  # 4
    print(sol.characterReplacement(s2, 1))  # 4


if __name__ == "__main__":
    main()
